gobblin-data-management/src/main/java/gobblin/data/management/copy/extractor/CloseableFsFileAwareInputStreamExtractor.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.data.management.copy.extractor;
+
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.source.extractor.extract.sftp.SftpLightWeightFileSystem;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+import org.apache.hadoop.fs.FileSystem;
+
+import com.google.common.io.Closer;
+
+
+/**
+ * Used instead of {@link FileAwareInputStreamExtractor} for {@link FileSystem}s that need be closed after use E.g
+ * {@link SftpLightWeightFileSystem}.
+ */
+public class CloseableFsFileAwareInputStreamExtractor extends FileAwareInputStreamExtractor {
+
+  private Closer closer = Closer.create();
+
+  public CloseableFsFileAwareInputStreamExtractor(FileSystem fs, Iterator<CopyableFile> filesIterator)
+      throws IOException {
+
+    super(fs, filesIterator);
+    closer.register(fs);
```

Comment:
We should be careful when closing `FileSystem` objects. Hadoop `FileSystem`s has this feature where it keeps an internal cache of `FileSystem` objects, so `FileSystem.get()` is usually returning a cached version of the `FileSystem`. When `FileSystem.close()` is called the cached `FileSystem is closed, which entails closing the connection to HDFS. Once the connection is closed on the cached `FileSystem` any other code using the cached version will fail. I've hit the issue at least once before. http://stackoverflow.com/questions/20057881/hadoop-filesystem-closed-exception-when-doing-bufferedreader-close has more info.